### DISCUSSION TITLE
Complex file and folder structures

### DISCUSSION THREAD

**Discussion by beriberikix** [codapi-org/support#2] [Q&A] [answered]
I'm trying to implement a sandbox for a C application that has very specific file and folder requirements. Specific files need to exist for compilation to work, they need to be in specific directories, etc. I'm struggling to figure out how exactly to implement the sandbox and would love some guidance. I think the problem is not unique to my application (a RTOS called https://www.zephyrproject.org/) and might be something others will run into.

I created a [branch](https://github.com/beriberikix/codapi/tree/zephyr) in my fork with how I would _expect_ it to work but ran into a series of issues. To summarize the high level requirements:

- specific files need to exist in the cwd and in specifically named folders when running the build step (ex. [these files](https://github.com/beriberikix/codapi/blob/zephyr/images/zephyr/Dockerfile#L8-L10))
  - These files are not present when running the command (maybe they're not in the entry point?)
- A precompile step needs to happen for every run and would ideally happen at the docker image generation phase
  - Like the specific files above, all the code in this step is not available when the sandbox is executed
- Setting the entrypoint to a subdirectory fails
  - zephyr_run_82c682c2: write files to temp dir: open /tmp/255357378/app/main.c: no such file or directory

I have a hunch that these may be related. It might be that I'm missing a key concept or this isn't a use case that's currently supported. But any advice or feedback would be greatly appreciated!

**Comment by nalgeon** [comment, marked as answer]
I suggest you get the sandbox working with just the Docker container and then add the Codapi layer.

Build the image:

```shell
build --file images/zephyr/Dockerfile --tag codapi/zephyr images/zephyr/
```

Check that the files are in place (they are):

```shell
docker run --rm codapi/zephyr ls /sandbox/app
```

```
CMakeLists.txt
prj.conf
src
west.yml
```

Check the build command (it works up to a point because there are no source files in `/sandbox/app/src`):

```
docker run --rm codapi/zephyr west build -b native_sim app/
```

```
Loading Zephyr default modules (Zephyr base).
-- Application: /sandbox/app
-- CMake version: 3.25.1
-- Found Python3: /opt/venv/bin/python3 (found suitable version "3.11.2", minimum required is "3.8") found components: Interpreter 
-- Cache files will be written to: /root/.cache/zephyr
-- Zephyr version: 3.6.99 (/sandbox/deps/zephyr)
...
CMake Error at /sandbox/deps/zephyr/cmake/modules/extensions.cmake:425 (add_library):
  No SOURCES given to target: app
```

Create `/sandbox/app/src/main.c` and try to build it (I created one in `/tmp/c` on the host machine):

```
docker run --rm --volume /tmp/c:/sandbox:ro codapi/zephyr west build -b native_sim app/
```

And now it fails:

```
usage: west [-h] [-z ZEPHYR_BASE] [-v] [-V] <command> ...
west: unknown command "build"; do you need to run this inside a workspace?
```

The reason is that when you mount the local directory (`/tmp/c` in my example) as the container directory (`/sandbox`), it replaces everything, including the `/sandbox/app` directory you created when building an image. This is how Docker volumes work, and is not specific to Codapi.

The solution is to map the host directory to `/sandbox/app/src` instead of `/sandbox`:

```
docker run --rm --volume /tmp/c:/sandbox/app/src:ro codapi/zephyr west build -b native_sim app/
```

Now we need to do the same in the Codapi `configs/boxes.json`:

```json
{
    "zephyr": {
        "image": "codapi/zephyr",
        "writable": true,
        "volume": "%s:/sandbox/app/src:ro"
    }
}
```

I've also set `writable` to `true`, because the build process seems to be writing something to `/sandbox` and I have no idea how to change this (you probably know better).

Finally, we can prepare the `configs/commands/zephyr.json`:

```json
{
    "run": {
        "engine": "docker",
        "entry": "main.c",
        "steps": [
            {
                "box": "zephyr",
                "user": "root",
                "timeout": 60,
                "command": ["west", "build", "-b", "native_sim", "app/"]
            }
        ]
    }
}
```

I set the `user` to `root` because in the Dockerfile you removed the `sandbox` user. I also had to set a really high `timeout` because the build process is very slow (which kind of destroys the whole idea of having an interactive sandbox for me, but it's up to you of course).

**Reply by beriberikix** [reply]
Thank you for the explanation, very helpful!

I'm still struggling with the permissions file writing permissions. The processes generates a few files during the build step so I don't think I can get around that.

Following your instructions I'm getting:

```bash
jberi@jberi-lenovo:~/code/codapi$ curl -H "content-type: application/json" -d '{ "sandbox": "zephyr", "command": "run", "files": {"main.c": "/* */"}}' http://localhost:1313/v1/exec | jq .
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  1345  100  1275  100    70   4162    228 --:--:-- --:--:-- --:--:--  4395
{
  "id": "zephyr_run_73504148",
  "ok": false,
  "duration": 305,
  "stdout": "",
  "stderr": "Traceback (most recent call last):\n  File \"/opt/venv/bin/west\", line 8, in <module>\n    sys.exit(main())\n             ^^^^^^\n  File \"/opt/venv/lib/python3.11/site-packages/west/app/main.py\", line 1085, in main\n    app.run(argv or sys.argv[1:])\n  File \"/opt/venv/lib/python3.11/site-packages/west/app/main.py\", line 244, in run\n    self.run_command(argv, early_args)\n  File \"/opt/venv/lib/python3.11/site-packages/west/app/main.py\", line 505, in run_command\n    self.run_extension(args.command, argv)\n  File \"/opt/venv/lib/python3.11/site-packages/west/app/main.py\", line 654, in run_extension\n    self.cmd.run(args, unknown, self.topdir, manifest=self.manifest,\n  File \"/opt/venv/lib/python3.11/site-packages/west/commands.py\", line 194, in run\n    self.do_run(args, unknown)\n  File \"/sandbox/deps/zephyr/scripts/west_commands/build.py\", line 193, in do_run\n    self._setup_build_dir()\n  File \"/sandbox/deps/zephyr/scripts/west_commands/build.py\", line 404, in _setup_build_dir\n    os.makedirs(build_dir, exist_ok=False)\n  File \"<frozen os>\", line 225, in makedirs\nOSError: [Errno 30] Read-only file system: '/sandbox/build' (exit status 1)"
}
```

I tried changing the boxes volume config to `"volume": "%s:/sandbox/app/src:rw"` as well as tried setting the global box config to `writeable` with the same result. Any ideas?

**Reply by nalgeon** [reply]
My comment above contains the working config, I've checked it. The "read-only file system" error message clearly indicates that `writable` is not set to `true.` Did you restart the application after changing the configs?

**Reply by beriberikix** [reply]
Yes, and even restarted my machine just in case there was some possible form of caching. I keep rechecking and trying it but still the same issue.
![image](https://github.com/codapi-org/support/assets/178376/b5ef571e-ee82-4157-afa8-7836050a1ea6)
I can only assume a typo, so I created a [clean branch](https://github.com/beriberikix/codapi/tree/zephyr-posix) with the changes I'm testing.

**Reply by nalgeon** [reply]
It's `writable`, not `writeable`.
